libdrgn/dwarf_info_cache.c

```diff
@@ -1475,6 +1478,50 @@ drgn_dwarf_object_find(const char *name, size_t name_len, const char *filename,
 	return &drgn_not_found;
 }
 
+struct drgn_error *
+drgn_dwarf_object_find_with_namespaces(const char *name, size_t name_len, const char *filename,
+				       enum drgn_find_object_flags flags,
+				       struct drgn_dwarf_index_namespace* ns,
+				       void *arg, struct drgn_object *ret) {
+	struct drgn_error *err;
+	struct drgn_dwarf_info_cache *dicache = arg;
+	char* ns_ptr;
+
+	if (!ns)
+		ns = &dicache->dindex.ns;
+
+find_ns:
+	ns_ptr = strstr(name, "::");
+	if (ns_ptr) {
+		const int ns_len = ns_ptr - name;
+		const size_t ns_colon_len = ns_len + 2;
+		struct drgn_dwarf_index_iterator it;
+		uint64_t ns_tag = DW_TAG_namespace;
+		struct drgn_dwarf_index_die *die;
+
+		drgn_dwarf_index_iterator_init(&it, ns, name, ns_len, &ns_tag, 1);
+		// Namespaces should be de-duplicated so we only check for the first one.
+		while ((die = drgn_dwarf_index_iterator_next_die(&it, true))) {
+			if (drgn_dwarf_index_die_namespace(die)) {
```

Comment:
Since we're only iterating over namespace DIEs here, we don't need to check this, right?

Reply:
In theory this should never be false but I was worried that if there was some bug where we returned a non-namespace type, this would explode, I'll remove it.